src/base/cub_ddl_log.c

```diff
@@ -0,0 +1,1054 @@
+cub_ddl_log_open (char *app_name)
+{
+  FILE *fp = NULL;
+  char *tpath = NULL;
+  int len;
+
+  if (app_name != NULL)
+    {
+      len = cub_make_ddl_log_filename (ddl_audit_handle->log_filepath, PATH_MAX, app_name);
+
+      if (ddl_audit_handle->log_filepath[0] == '\0' || len < 0)
+	{
+	  goto file_error;
+	}
+
+      tpath = strdup (ddl_audit_handle->log_filepath);
+      if (tpath == NULL)
+	{
+	  goto file_error;
+	}
+      cub_create_dir_log (tpath);
+      FREE_MEM (tpath);
+    }
+  else
+    {
+      goto file_error;
+    }
+
+  /* note: in "a+" mode, output is always appended */
+  fp = cub_fopen_and_lock (ddl_audit_handle->log_filepath, "a+");
+
+  if (fp != NULL)
+    {
+      fseek (fp, 0, SEEK_END);
+    }
+  return fp;
+
+file_error:
+  return NULL;
+}
+
+void
+cub_ddl_log_write_end ()
+{
+  if (ddl_audit_handle == NULL)
+    {
+      return;
+    }
+
+  if (prm_get_bool_value (PRM_ID_DDL_AUDIT_LOG) == FALSE)
+    {
+      goto ddl_log_free;
+    }
+
+  if (strcmp (ddl_audit_handle->app_name, "loaddb") == 0)
+    {
+      if (ddl_audit_handle->loaddb_file_type != LOADDB_FILE_TYPE_SCHEMA)
+	{
+	  goto ddl_log_free;
+	}
+    }
+  else
+    {
+      if (cub_is_ddl_type (ddl_audit_handle->stmt_type) == FALSE)
+	{
+	  goto ddl_log_free;
+	}
+    }
+
+  if (ddl_audit_handle->log_type & DDL_LOG_ELAPSED_TIME)
+    {
+      cub_ddl_log_write ();
+    }
+
+ddl_log_free:
+  cub_ddl_log_free ();
+}
+
+void
+cub_ddl_log_write ()
+{
+  FILE *fp = NULL;
+  char buf[DDL_LOG_BUFFER_SIZE] = { 0 };
+  char dest_path[PATH_MAX] = { 0 };
+  int len = 0;
+  int ret = 0;
+
+  if (ddl_audit_handle == NULL)
+    {
+      return;
+    }
+
+  if (prm_get_bool_value (PRM_ID_DDL_AUDIT_LOG) == FALSE)
+    {
+      return;
+    }
+
+  if (strcmp (ddl_audit_handle->app_name, "loaddb") == 0)
+    {
+      if (ddl_audit_handle->loaddb_file_type != LOADDB_FILE_TYPE_SCHEMA)
+	{
+	  return;
+	}
+    }
+  else
+    {
+      if (cub_is_ddl_type (ddl_audit_handle->stmt_type) == FALSE)
+	{
+	  return;
+	}
+    }
+
+  fp = cub_ddl_log_open (ddl_audit_handle->app_name);
+
+  if (fp != NULL)
+    {
+      if (strcmp (ddl_audit_handle->app_name, "loaddb") == 0)
+	{
+	  if (cub_make_schema_file_name (ddl_audit_handle->file_name, dest_path, PATH_MAX) < 0)
+	    {
+	      goto write_error;
+	    }
+	}
+
+      len = cub_create_log_mgs (buf);
+      if (len < 0)
+	{
+	  goto write_error;
+	}
+
+      if (fwrite (buf, 1, len, fp) != len)
+	{
+	  goto write_error;
+	}
+
+      if (strcmp (ddl_audit_handle->app_name, "loaddb") == 0)
+	{
+	  cub_file_copy (ddl_audit_handle->file_name, dest_path);
+	}
+
+      if ((UINT64) ftell (fp) > prm_get_bigint_value (PRM_ID_DDL_AUDIT_LOG_SIZE))
+	{
+	  fclose (fp);
+	  cub_ddl_log_backup (ddl_audit_handle->log_filepath);
+	  fp = NULL;
+	}
+    }
+
+write_error:
+  if (fp != NULL)
+    {
+      fclose (fp);
+      fp = NULL;
+    }
+}
+
+static int
+cub_create_log_mgs (char *msg)
+{
+  int retval = 0;
+  char result[20] = { 0 };
+
```

Comment:
Minor suggestion:
```
if (ddl_audit_handle == NULL)
  {
     return -1;
  }
```

Reply:
Thank you, i changed the code which you commented.